### ISSUE TITLE
SameDiff: gradient calculation wrong after updating 

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#5680]
```
    @Test
    public void testUpdatingGradientSimple(){

        SameDiff sd = SameDiff.create();
        SDVariable in = sd.var("in", Nd4j.linspace(1,12,12).reshape(3,4));
        SDVariable out = in.mul(2.0);
        SDVariable loss = out.std("out", true);

        INDArray outArr = sd.execAndEndResult().dup();
        sd.execBackwards();

        Map<String,INDArray> origGrad = new HashMap<>();
        origGrad.put("in", in.gradient().getArr().dup());
        origGrad.put("out", out.gradient().getArr().dup());

        double stdBefore = in.getArr().stdNumber().doubleValue();
        in.getArr().muli(10);
        double stdAfter = in.getArr().stdNumber().doubleValue();
        System.out.println("Before vs. after: " + stdBefore + ", " + stdAfter);
        INDArray outArr2 = sd.execAndEndResult();
        sd.execBackwards();

        assertNotEquals(outArr, outArr2);

        //Ensure gradients are also changed:
        assertNotEquals(origGrad.get("in"), in.gradient().getArr());        //Fails here: same gradient as before updating the array
        assertNotEquals(origGrad.get("out"), out.gradient().getArr());
    }
```

**Comment by AlexDBlack** [comment]
Hm... turns out that stdev gradient cancels out the muli(10) here. Assign random values instead and it passes.
So this isn't the root cause of another bug that I'm seeing...
